### PR TITLE
Change SelectionResult structure, add bool checks.

### LINKED ISSUE #336: Bool values aren't enforced in SelectEvents and SelectionResult objects
While playing around with columnflow, I stumbled upon an undesired behavior that could happen particularly to new users. In my setup, I made a logical "error". Specifically, I created a new `Selector` and have this line when creating an initial selection step that selects all events

```python
# do some selections which yield SelectionResult `result`
result.steps.all = ak.full_like(events.event, True)
```

The error I made is assuming that this would yield an array of bools, where it actually yields an array of type `uint64`. So far, this is of course unrelated to columnflow.

However, this small "error" lead to the undesired behavior: all subsequent reduction steps (e.g. `and` operator of different selection steps) themselves yield `uint64`. This doesn't throw any errors, but in the `ReduceEvents` tasks this arrays are then interpretated as **indices**, not masks of `bool`s. This is of course absolute nonsense since you effectively use only two events multiple times. However, it is very hard/annoying to debug this since there aren't any errors.

It would help immensely if we'd implement an explicit check for the type of arrays that are not located in `objects`, since these should always be of type `bool`. This check could in principle be done in two places imho

- when casting a `SelectionResult` into an awkward array
- when writing awkward arrays to parquet at the end of the `SelectEvents` task

Additionally, an explicite type check in the `ReduceEvents` task when creating the final `event_mask` [here](https://github.com/columnflow/columnflow/blob/master/columnflow/tasks/reduction.py#L193-L198) would also help avoid this issue in the future.

### PR DESCRIPTION
**Pull request by riga**
This PR slightly restructures the contents of the `SelectionResult` and adds a couple sanity checks as documented in #336 and #337.

Changes to `SelectionResult`:

- The main event mask used to be stored under `result.main.event`. This is now simply `result.event`.
- Other top-level fields (besides objects and steps) were also stored under `result.main.*` before but are now moved to `result.*` as well.

Added checks:

- When converting to arrays via the `to_ak()` method, there is now a check in place that verifies that event masks are booleans. We strictly require that to preserve the order of events across columns written in different places throughout columnflow.
- Inside `SelectEvents` an error is raised now in case the `selector_inst` returns a `SelectionResult` that has no event mask set. This is not entirely necessary per se, but we have some producers and other helpers that access the event mask, and it is cleaner to force the mask being there (which it is anyway in 99% of the cases), instead of having to do conditional checks in N places.

Fixes #336 and #337.

### CODE REVIEW

**Comment by riga** [comment]
Also adding @mafrahm , @JulesVandenbroeck and @gsaha009 here since this might require a change on their end.

**Review by pkausw — CHANGES_REQUESTED** [review]
Nice PR, thanks @riga! Just some minor changes and clarifications :+1:

File: columnflow/selection/__init__.py [hunk 1]
```diff
@@ -287,12 +295,22 @@ def to_ak(self) -> ak.Array:
...
-        The conversion is performed with multiple calls of
-        :external+ak:py:func:`ak.zip`.
+        The conversion is performed with multiple calls of :external+ak:py:func:`ak.zip`.
 
-        :return: Transformed :py:class:`~.SelectionResult`
+        :raises ValueError: If the main events mask contains a type other than bool.
+        :return: :py:class:`~.SelectionResult` transformed into an awkward array.
         """
+        # complain if the event mask consists of non-boolean values
+        if self.event is not None and getattr(ak.type(self.event).content, "primitive", None) != "bool":
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/selection/__init__.py:304]: Yup, the "primitive" attribute is a actually a string, so the comparison here should be independent of the underlying type.

File: columnflow/selection/__init__.py [hunk 2]
```diff
@@ -287,12 +295,22 @@ def to_ak(self) -> ak.Array:
         """
         Converts the contained fields into a nested awkward array and returns it.
 
-        The conversion is performed with multiple calls of
-        :external+ak:py:func:`ak.zip`.
+        The conversion is performed with multiple calls of :external+ak:py:func:`ak.zip`.
 
-        :return: Transformed :py:class:`~.SelectionResult`
+        :raises ValueError: If the main events mask contains a type other than bool.
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/selection/__init__.py:300]: Added

File: columnflow/selection/__init__.py [hunk 3]
```diff
@@ -208,36 +210,38 @@ class SelectionResult(od.AuxDataMixin):
                 "Muon": {
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/selection/__init__.py:213]: I rephrased the comments slightly to be in line with the comments above that block.

**Review by pkausw — APPROVED** [review]
LGTM :+1:

**Comment by JulesVandenbroeck** [comment]
Hi, I tried to adapt the code after pulling from master. However afterwards It seems I can no longer update SelectionResult objects (aka: _results += lepton_results_) due to the error message "*** TypeError: cannot pickle 'dict_keys' object". As I am new to columnflow I am mainly following the example of the hh2bbww analysis (which has not yet pushed this change). In case the answer is not obvious, I linked the [gitlab of my project](https://gitlab.cern.ch/juvanden/ttz) (selection is under ttz/selection/default.py")

**Pull request merged**